### PR TITLE
Static instance of OR collector

### PR DESCRIPTION
**Pull request by ppiotrow**
Small changes while getting familiar with the new Parallel executor.

### CODE REVIEW

**Comment by ppiotrow** [comment]
Coverage warning is false positive this time.

File: roaringbitmap/src/main/java/org/roaringbitmap/ParallelAggregation.java [hunk 1]
```diff
@@ -139,12 +141,8 @@ public ContainerCollector(Function<List<Container>, Container> reducer) {
...
       for (int i = 0; i < ra.size; ++i) {
         Container container = ra.values[i];
-        Short key = ra.keys[i];
-        List<Container> slice = grouped.get(key);
-        if (null == slice) {
-          slice = new ArrayList<>();
-          grouped.put(key, slice);
-        }
+        short key = ra.keys[i];
+        List<Container> slice = grouped.computeIfAbsent(key, k -> new ArrayList<>());
```
Comment (richardstartin) [inline roaringbitmap/src/main/java/org/roaringbitmap/ParallelAggregation.java]: Please benchmark this. The code was written this way for a reason.

File: roaringbitmap/src/main/java/org/roaringbitmap/buffer/BufferParallelAggregation.java [hunk 2]
```diff
@@ -144,12 +146,8 @@ public ContainerCollector(Function<List<MappeableContainer>, MappeableContainer>
       MappeableContainerPointer it = bitmap.highLowContainer.getContainerPointer();
       while (null != it.getContainer()) {
         MappeableContainer container = it.getContainer();
-        Short key = it.key();
```
Comment (richardstartin) [inline roaringbitmap/src/main/java/org/roaringbitmap/buffer/BufferParallelAggregation.java]: Likewise.

**Comment by richardstartin** [comment]
I _insist_ on seeing evidence that the change to `groupByKey` doesn't degrade performance. There is a benchmark for the method. My numbers with and without your change indicate that it hurts most when you would already be feeling the cost of the group by.

minor-changes-parallel:


Benchmark | (dataset) | Mode | Cnt | Score | Error | Units
-- | -- | -- | -- | -- | -- | --
ParallelAggregatorBenchmark.groupByKey | census-income | avgt | 5 | 15.935 | 3.828 | us/op
ParallelAggregatorBenchmark.groupByKey | census1881 | avgt | 5 | 35.125 | 0.128 | us/op
ParallelAggregatorBenchmark.groupByKey | dimension_008 | avgt | 5 | 207.335 | 3.04 | us/op
ParallelAggregatorBenchmark.groupByKey | dimension_003 | avgt | 5 | 520.126 | 243.285 | us/op
ParallelAggregatorBenchmark.groupByKey | dimension_033 | avgt | 5 | 40.467 | 7.312 | us/op
ParallelAggregatorBenchmark.groupByKey | uscensus2000 | avgt | 5 | 111.102 | 3.596 | us/op
ParallelAggregatorBenchmark.groupByKey | weather_sept_85 | avgt | 5 | 52.822 | 1.182 | us/op
ParallelAggregatorBenchmark.groupByKey | wikileaks-noquotes | avgt | 5 | 38.374 | 1.226 | us/op
ParallelAggregatorBenchmark.groupByKey | census-income_srt | avgt | 5 | 13.098 | 1.089 | us/op
ParallelAggregatorBenchmark.groupByKey | census1881_srt | avgt | 5 | 54.319 | 0.48 | us/op
ParallelAggregatorBenchmark.groupByKey | weather_sept_85_srt | avgt | 5 | 42.935 | 0.494 | us/op
ParallelAggregatorBenchmark.groupByKey | wikileaks-noquotes_srt | avgt | 5 | 32.624 | 0.252 | us/op

Master:


Benchmark | (dataset) | Mode | Cnt | Score | Error | Units
-- | -- | -- | -- | -- | -- | --
ParallelAggregatorBenchmark.groupByKey | census-income | avgt | 5 | 14.393 | 1.03 | us/op
ParallelAggregatorBenchmark.groupByKey | census1881 | avgt | 5 | 33.636 | 0.689 | us/op
ParallelAggregatorBenchmark.groupByKey | dimension_008 | avgt | 5 | 190.664 | 11.762 | us/op
ParallelAggregatorBenchmark.groupByKey | dimension_003 | avgt | 5 | 422.888 | 145.04 | us/op
ParallelAggregatorBenchmark.groupByKey | dimension_033 | avgt | 5 | 38.417 | 4.705 | us/op
ParallelAggregatorBenchmark.groupByKey | uscensus2000 | avgt | 5 | 95.83 | 2.322 | us/op
ParallelAggregatorBenchmark.groupByKey | weather_sept_85 | avgt | 5 | 51.077 | 0.751 | us/op
ParallelAggregatorBenchmark.groupByKey | wikileaks-noquotes | avgt | 5 | 36.751 | 0.433 | us/op
ParallelAggregatorBenchmark.groupByKey | census-income_srt | avgt | 5 | 12.916 | 0.137 | us/op
ParallelAggregatorBenchmark.groupByKey | census1881_srt | avgt | 5 | 50.593 | 0.344 | us/op
ParallelAggregatorBenchmark.groupByKey | weather_sept_85_srt | avgt | 5 | 39.133 | 0.404 | us/op
ParallelAggregatorBenchmark.groupByKey | wikileaks-noquotes_srt | avgt | 5 | 31.781 | 0.395 | us/op

File: roaringbitmap/src/main/java/org/roaringbitmap/ParallelAggregation.java (hunk 1 above)
Comment (ppiotrow) [reply roaringbitmap/src/main/java/org/roaringbitmap/ParallelAggregation.java]: The benchmarks proved that you're right. 
I wonder why just even change `Short -> short` has small degradation effect on all benchmarks. Do you know the answer?

Comment (richardstartin) [reply roaringbitmap/src/main/java/org/roaringbitmap/ParallelAggregation.java]: Autoboxing.

**Comment by ppiotrow** [comment]
Only having OR collector as static instance gives some performance improvement. Other changes were causing degradation of the performance. 
```
master
Benchmark                                            (dataset)  Mode  Cnt     Score    Error  Units
ParallelAggregatorBenchmark.parallelOr           census-income  avgt    5   205.300 ± 57.328  us/op
ParallelAggregatorBenchmark.parallelOr              census1881  avgt    5   344.132 ±  2.794  us/op
ParallelAggregatorBenchmark.parallelOr           dimension_008  avgt    5   674.686 ±  8.366  us/op
ParallelAggregatorBenchmark.parallelOr           dimension_003  avgt    5  1521.433 ± 28.428  us/op
ParallelAggregatorBenchmark.parallelOr           dimension_033  avgt    5   502.853 ±  6.350  us/op
ParallelAggregatorBenchmark.parallelOr            uscensus2000  avgt    5   132.675 ±  3.848  us/op
ParallelAggregatorBenchmark.parallelOr         weather_sept_85  avgt    5   838.342 ± 30.714  us/op
ParallelAggregatorBenchmark.parallelOr      wikileaks-noquotes  avgt    5   152.254 ±  1.006  us/op
ParallelAggregatorBenchmark.parallelOr       census-income_srt  avgt    5   185.338 ±  3.299  us/op
ParallelAggregatorBenchmark.parallelOr          census1881_srt  avgt    5   263.366 ±  7.902  us/op
ParallelAggregatorBenchmark.parallelOr     weather_sept_85_srt  avgt    5   408.157 ± 12.176  us/op
ParallelAggregatorBenchmark.parallelOr  wikileaks-noquotes_srt  avgt    5    98.757 ±  2.746  us/op

static instance of OR collector
Benchmark                                            (dataset)  Mode  Cnt     Score    Error  Units
ParallelAggregatorBenchmark.parallelOr           census-income  avgt    5   180.659 ± 68.770  us/op
ParallelAggregatorBenchmark.parallelOr              census1881  avgt    5   333.633 ±  2.701  us/op
ParallelAggregatorBenchmark.parallelOr           dimension_008  avgt    5   666.207 ± 13.730  us/op
ParallelAggregatorBenchmark.parallelOr           dimension_003  avgt    5  1417.407 ± 51.410  us/op
ParallelAggregatorBenchmark.parallelOr           dimension_033  avgt    5   500.968 ±  7.236  us/op
ParallelAggregatorBenchmark.parallelOr            uscensus2000  avgt    5   128.742 ±  5.457  us/op
ParallelAggregatorBenchmark.parallelOr         weather_sept_85  avgt    5   768.676 ± 32.145  us/op
ParallelAggregatorBenchmark.parallelOr      wikileaks-noquotes  avgt    5   148.195 ±  3.143  us/op
ParallelAggregatorBenchmark.parallelOr       census-income_srt  avgt    5   186.608 ±  3.326  us/op
ParallelAggregatorBenchmark.parallelOr          census1881_srt  avgt    5   263.444 ±  6.084  us/op
ParallelAggregatorBenchmark.parallelOr     weather_sept_85_srt  avgt    5   409.154 ± 16.161  us/op
ParallelAggregatorBenchmark.parallelOr  wikileaks-noquotes_srt  avgt    5    98.988 ±  1.243  us/op
```
BTW @richardstartin , how do you generate this nice tables with benchmark comparison?

File: roaringbitmap/src/main/java/org/roaringbitmap/ParallelAggregation.java (hunk 1 above)
Comment (richardstartin) [reply roaringbitmap/src/main/java/org/roaringbitmap/ParallelAggregation.java]: Moreover, why edit for the hell of it, if you abstain from all welcome code review?

Comment (ppiotrow) [reply roaringbitmap/src/main/java/org/roaringbitmap/ParallelAggregation.java]: Yup, auto boxing as it is later used as a key in collection, didn't notice that.
D

**Comment by richardstartin** [comment]
I think you should rebase and submit a single commit containing a single change: make the `OrCollector` instance static.

For the tables I use github markdown.

If you are very keen to make this faster, one idea to try is to make it work with just a `HashMap`. You can implement a specialised radix sort which reorders the values as it shuffles the keys after the aggregation is complete. This will halve the time spent in `groupByKey` and reduce GC pressure. The final reordering is linear time and cheap.

**Comment by lemire** [comment]
@ppiotrow Can you answer @richardstartin's comment?

**Comment by ppiotrow** [comment]
It can be squashed while merging via "squash and merge".
Regarding markdown I though that maybe you have script that does it for you.

The radix sort improvement, maybe just add it as an issue, as I can get involved right now.

**Comment by lemire** [comment]
I'll merge.

**Comment by lemire** [comment]
Merged.

**Pull request merged**